Add rule vue/no-early-return

jesusgn90 opened the pull request:
fixes https://github.com/vuejs/eslint-plugin-vue/issues/1884

FloEdelmann commented on lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,76 @@
+          for (const loc of rStatementLocs) {
+            context.report({
+              loc,
+              message: 'Extra return statement in setup function.'
```
Please change this message to:

```suggestion
              message: 'Only one (unconditional) return statement is allowed in setup function.'
```

Even better: Create a shared message with a `messageId`: https://eslint.org/docs/developer-guide/working-with-rules#messageids

```js
messages: {
  extraReturnStatement: 'Only one (unconditional) return statement is allowed in {{ functionType }} function.'
}
```

Then reuse this message here:

```js
messageId: 'extraReturnStatement',
data: {
  functionType: 'setup'
},
```

FloEdelmann commented on lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,76 @@
+          ) {
+            context.report({
+              node,
+              message: 'Extra return statement in data function.'
```
```suggestion
              message: 'Only one (unconditional) return statement is allowed in data function.'
```

FloEdelmann commented on lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,76 @@
+
+  /** @param {RuleContext} context */
+  create(context) {
+    const rStatementLocs = []
```
Please rename this function to `returnStatementLocations`.

FloEdelmann commented on lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,76 @@
+          }
+        },
+        onVueObjectEnter(node) {
+          const dataProperty = utils.findProperty(node, 'data')
```
The same logic should also apply to Nuxt.js's `asyncData` property: https://nuxtjs.org/docs/directory-structure/pages#asyncdata

FloEdelmann commented on lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,76 @@
+  /** @param {RuleContext} context */
+  create(context) {
+    const rStatementLocs = []
+    return utils.compositingVisitors(
```
Is this needed? Can't you return `utils.defineVueVisitor(…)` directly?

FloEdelmann reviewed (changes requested):
Wow, that was fast, thanks! I have a few remarks.

FloEdelmann commented on docs/rules/no-early-return.md:
```diff
@@ -0,0 +1,103 @@
+
+## :book: Rule Details
+
+This rule helps to identify accidental `return` statements in the `setup` block of a Vue component, these are `return` statements that would exit the `setup` block skipping part of the block itself (not allowing to each the end of the block).
+
+On the other hand, the rule checks the `data` block too, applying the same logic.
```
```suggestion
The `setup` and `data` functions of Vue components (and [`asyncData`](https://nuxtjs.org/docs/directory-structure/pages/#asyncdata) of Nuxt.js page components) should only have a single (unconditional) return statement at the end. This rule reports early `return` statements, i.e. those that would exit the function early without allowing to reach the end of the function body.
```

jesusgn90 replied:
applied this without the part that mentions Nuxt

jesusgn90 commented:
@FloEdelmann thanks for your review, I think I've addressed your comments. There is one point however, the one related to Nuxt that I'm not super sure so I'd prefer to leave the rule as it is for now. Perhaps, it's worth opening a new issue specific for that so others can comment on making this rule cover that for Nuxt or not. For me, even covering the `data` part was an extra, the original purpose was to cover just extra returns on `setup` blocks :smile:

FloEdelmann commented:
The logic for `asyncData` would be exactly the same as for `data`.

jesusgn90 commented:
> The logic for asyncData would be exactly the same as for data.

@FloEdelmann ok, if they work the same, I just pushed a small change to cover it (and in a way that if we need to add more properties in the future, it's easily extendable)

FloEdelmann commented on tests/lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,302 @@
+        }
+      ]
+    },
+    {
```
Please add one more test case. I'm afraid this will produce too many errors :slightly_frowning_face: 

```suggestion
    {
      filename: 'test.vue',
      code: `
      <script>
      export default {
        setup: () => {
          if (maybe) {
            return { foo: false }
          }
          return { foo: true }
        },
        data() {
          return {
            foo: true
          }
        },
        computed: {
          foo() {
            return true
          }
        }
      }
      </script>
      `,
      errors: [
        {
          messageId: 'extraReturnStatement',
          data: { functionType: 'setup' },
          line: 3
        }
      ]
    },
    {
```

ota-meshi replied:
I found a false positive in the next test source.

```vue
<script>
export default {
  methods: {
    fn () {
      if (a) {
        return {}; // <- false positive
      }
    }
  },
  setup () {
    const foo = ref()
    return { foo }
  }
}
</script>
```

I think we need to check the scope of ReturnStatement correctly.
The implementation of the return-in-emits-validator rule may be helpful.
https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/return-in-emits-validator.js

FloEdelmann commented on tests/lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,302 @@
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        setup () {
+          const foo = computed(() => {
+            if (s) {
+              return
+            }
+          })
+        }
+      }
+      </script>
+      `
+    }
```
Unfortunately, this test case fails although it should pass:

https://github.com/vuejs/eslint-plugin-vue/runs/6426396807?check_suite_focus=true#step:5:187

FloEdelmann commented on lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,85 @@
+        returnStatementLocations.push(node.loc)
+      },
+
+      onSetupFunctionExit() {
```
Can't the logic for `data`/`asyncData` and for `setup` be shared? I think they should work/behave exactly the same.

ota-meshi commented on tests/lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,302 @@
+        {
+          messageId: 'extraReturnStatement',
+          data: { functionType: 'data' },
+          line: 3
```
I think the report should point to line 6.

ota-meshi commented on tests/lib/rules/no-early-return.js:
```diff
@@ -0,0 +1,302 @@
+      </script>
+      `
+    }
+  ],
```
I found a false positive in the next test source.

```vue
<script>
export default {
  methods: {
    fn () {
      if (a) {
        return {}; // <- false positive
      }
    }
  },
  setup () {
    const foo = ref()
    return { foo }
  }
}
</script>
```

I think we need to check the scope of ReturnStatement correctly.
The implementation of the return-in-emits-validator rule may be helpful.
https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/return-in-emits-validator.js

ota-meshi replied:
Or I think you can use code path analysis like `consistent-return` rule.
https://eslint.org/docs/developer-guide/code-path-analysis
https://github.com/eslint/eslint/blob/main/lib/rules/consistent-return.js

jesusgn90 commented:
I'll resume this with more time
